SEP-0008: Add additional optional action params for the action required response (#758)
stellar/stellar-protocol: ecosystem/sep-0008.md

#### Action Required Response

> An Action Required response will have a `200` HTTP status code and `action_required` as the `status` value. It means that the user must complete an action before this transaction can be approved. The approval service will provide a URL that facilitates the action. Upon completion, the user will resubmit the transaction if it's synchronous (the default) or wait for a confirmation from the approval service to continue.

Reviewer: What does the waiting process look like? Does a client wait for some signal? What is that signal?

Author: In our case, the user get's an email that she has passed kyc and then she can trade. The client is not involved here. It basically means that afterwards the client does nothing.

Reviewer: Got it, so the app if it is synchronous would return to resending the same request again and retrying after the action required step, where-as if it is not synchronous the app would stop. Is that an accurate understanding?

Author: That's up to the implementation, I'm not sure how current clients handle that, but if it's synchronous, then yes, the client could do that. 

Maybe we can leave that out for now as well, I'm not aware of other SEP-8 implementations so I'm not sure if that is actually a thing.

status|string|`"action_required"`
message|string|A human readable string containing information regarding the action required.

> action_url|string|A URL that allows the user to complete the actions required to have the transaction approved.

Reviewer: Today it's assumed the client will open the URL in a browser, but providing these options might need to be POSTed to keep sensitive information out of the URL query params. We probably need to specify what request format this URL should be in the case actions params are provided.

Author: Yeah, i thought that the client would add those to the query params, as it's currently exactly like that (e.g. as you can see from our sample we are already adding the public key, taken from the tx.

Reviewer: I don't see a problem with adding the public key to the URL but the name and email could be leaked in other places by including it there. URLs are not always secure as they get easily propagated to logs, ad tracking, analytics systems, etc, so it would be ideal to keep these details in a POST either by submitting them in the request body as a POST separately and getting back a URL with an ID referencing that data, or by using a POST at this point although that might make it harder to implement in some cases such as mobile.

SEP-12 would help provide that parallel delivery of data but it sounds like that would be significantly different to what you're doing today. I think we should probably explore that. Return of a list of data fields needed could be delivered separately as the custom, and then the action URL is loaded for the user. Prior to SEP-12 the flow would probably need SEP-10 to authenticate as the account first which could be seamless.

Author: Those are valid points, especially with the POST request.

SEP-12, as I understand it, has only `NEEDS_INFO`, which isn't the case here. We can work without E-Mail by collecting it for ourself (not all clients may have that info at hand), so it's optional.

Reviewer: Got it, so what you're saying is if you don't provide this information then it would be okay, the interactive browser page would then ask for it so it isn't the same thing because this is prefilling not requiring data, and if all the data is prefilled it can be skipped, and if there are fields still required the page opened can collect it. Is that accurate?

Reviewer: @shredding @dmgmyza and I discussed this offline and an alternative approach where:

Add a new optional response field, `action_method`. The field has three valid states:
- Not provided, interpreted the same as if it is given the value `GET`
- `GET` – the action URL expects to receive a GET request, any fields requested may be passed as query parameters
- `POST` – the action URL expects to receive a POST request, any fields requested may be passed url-form-encoded or as JSON

Action URLs providing a `POST` method should response to the POST with a HTTP response with status code 301 and a `Location` header that is a URL where the user can complete the required actions with all of the POST data pre-filled or already accepted.

This approach supports products entirely in a web browser because the browser will follow the 301 redirect and will prevent odd UI where a user for some reason refreshes the page, they won't see warnings about they are refreshing a page that is accepting post data.

This approach supports products that are mobile apps loading the URL in a system browser because the app can make the POST request internally, read the location out of the response and provide it to a browser to load.

This approach is additive, any existing implementation would see no change. Any existing implementations that are providing optional fields/params as query parameters would be compatible. Any new implementations that plan to provide personal data or sensitive data could choose to use the POST method.

This approach does put some burden on the client to support different approaches though which is not great. We could encourage the POST approach for new implementations.

> action_params|string|A comma separated list of additional parameters that the client may optionally provide to the approval service; options are `first_name`, `last_name`, `email` and `public_key`

Reviewer: Are these params query parameters or fields to be POSTed to the URL?

Given this list of params is likely to expand, and we might need to specify descriptions for each, can we start a table that lists the params and reference that table here?

Or, if this list is likely to overlap significantly with [SEP-9](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0009.md), I think we should reference that SEP and support any of those fields, using the field names there for consistency.

We could also probably transition towards using [SEP-12](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0012.md) since that SEP provides greater control for a client to maintain and delete personal information, although that might be a breaking change and something we'd need to work towards.

Author: I thought about that as well, in our case it would be overhead. Yeah, the naming should be consistent with SEP-9; for us SEP-12 does not make sense as we have our own KYC Flow.

Reviewer: Okay got it. If it works for you, let's change this to reference SEP-9 and say any parameters there may be requested as params. Also, let's refer to them as fields because that's what SEP-9 refers to them as.
```suggestion
action_fields|string| (Optional) A comma separated list of additional fields defined by [SEP-9](sep-0009.md) that the client may optionally provide to the approval service.
```

- In the case that `action_method` is `POST` any fields requested may be passed as JSON in the request body, using content type `application/json`. The server should respond to the `POST` with a HTTP response with status code `200`, content-type `application/json` and a body containing the next URL that the user should be taken to.

   The next URL in the POST case is necessary to support products that are websites and products that are mobile applications. Products that run entirely in the browser can redirect the user to the next URL and the redirect prevents odd user experiences where a user refreshes a page and is prompted to re-post the data. Products that are mobile applications cannot open a system browser with a POST request. They can make the POST request internally, extract the redirect in the response, and open a system browser with the next URL.

   Parameters:

   Name | Type | Description
   -----|------|------------
   `next_url` | string | A URL where the user can complete the required actions with all the parameters included in the original POST pre-filled or already accepted.

   Example:

   ```json
   {
     "next_url": "https://..."
   }

>    ```

Reviewer: This section provides clear instruction for what the action_url's response message should contain, but is not clear about how the request should be formatted or constructed. There are a couple different approaches that an implementer could take, either using SEP-12 for inspiration, or using each SEP-9 field as a key in a flat object, or using each SEP-9 field as a hierarchy.

@shredding @dmgmyza Could you share how the request is formatted in your implementation?

I'll update this PR with the additional details once I hear from you.

Reviewer: I've opened PR https://github.com/DSTOQ/stellar-protocol/pull/3 that contains additional details.

`status` | string | `"action_required"`
`message` | string | A human readable string containing information regarding the action required.
`action_url` | string | A URL that allows the user to complete the actions required to have the transaction approved.

> `action_method` | string | (Optional) `GET` or `POST`, indicating the type of request that should be made to the `action_url`. If not provided, `GET` is assumed.

Reviewer 2: nit: I think in other places of this doc, we use `(optional)` instead of `(Optional)`. Maybe it's worth making it consistent.

Reviewer: https://github.com/DSTOQ/stellar-protocol/pull/4

`action_fields` | string[] | (Optional) An array of additional fields defined by [SEP-9 Standard KYC / AML fields] that the client may optionally provide to the approval service when sending the request to the `action_url`.

###### Example

```json
{
  "status": "action_required",
  "message": "Please sign the terms of service via the provided URL.",
  "action_url": "https://...",
  "action_method": "POST",
  "action_fields": ["email_address", "mobile_number"]
}
```

###### Action URL

> If the `action_method` is provided the client uses the type of request indicated when sending a request to the `action_url`. If `action_fields` are to be provided and the fields contain sensitive or personal information the `POST` case is preferred to ensure sensitive information is not transmitted in the URL.

Reviewer 2: Adding a comma helps me understand it better.

```suggestion
If the `action_method` is provided, the client uses the type of request indicated when sending a request to the `action_url`. If `action_fields` are to be provided and the fields contain sensitive or personal information, the `POST` case is preferred to ensure sensitive information is not transmitted in the URL.
```

Reviewer: https://github.com/DSTOQ/stellar-protocol/pull/4

> - In the case that `action_method` is `GET` any fields requested may be passed as query parameters in the `action_url`.

Reviewer 2: ```suggestion
- In the case that `action_method` is `GET`, any fields requested may be passed as query parameters in the `action_url`.
```

Reviewer: https://github.com/DSTOQ/stellar-protocol/pull/4

> - In the case that `action_method` is `POST` any fields requested may be passed as JSON in the request body, using content type `application/json`. The server should respond to the `POST` with a HTTP response with status code `200`, content-type `application/json` and a body containing the next URL that the user should be taken to.

Reviewer 2: ```suggestion
- In the case that `action_method` is `POST`, any fields requested may be passed as JSON in the request body using content type `application/json`. The server should respond to the `POST` with a HTTP response with status code `200`, content-type `application/json` and a body containing the next URL that the user should be taken to.
```

Reviewer: https://github.com/DSTOQ/stellar-protocol/pull/4

   The next URL in the POST case is necessary to support products that are websites and products that are mobile applications. Products that run entirely in the browser can redirect the user to the next URL and the redirect prevents odd user experiences where a user refreshes a page and is prompted to re-post the data. Products that are mobile applications cannot open a system browser with a POST request. They can make the POST request internally, extract the redirect in the response, and open a system browser with the next URL.

   Request Parameters:

   Any fields defined by [SEP-9 Standard KYC / AML fields].

   Request Example:

   ```json
   {
     "email_address": "user@example.com"
   }

>    ```

Reviewer: Note: This portion discussing the request format does not discuss how to send SEP-9 fields that have sub-fields. I'm intentionally leaving that undefined until I get an understanding of how SEP-12 works, so that I can make SEP-8 consistent with SEP-12. Related discussion: https://github.com/stellar/stellar-protocol/issues/832
